observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java

```diff
@@ -0,0 +1,436 @@
+          .setEventLogger(eventLogger)
+          .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
+          .setRpcId(rpcId);
+      logger.log(Level.FINEST, "Writing Half Close event to GcpLogSink");
+      sink.write(logEntryBuilder.build());
+    }
+
+    /**
+     * Logs cancellation.
+     */
+    void logCancel(
+        long seqId,
+        String serviceName,
+        String methodName,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+
+      GrpcLogRecord.Builder logEntryBuilder = createTimestamp()
+          .setSequenceId(seqId)
+          .setServiceName(serviceName)
+          .setMethodName(methodName)
+          .setEventType(EventType.GRPC_CALL_CANCEL)
+          .setEventLogger(eventLogger)
+          .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
+          .setRpcId(rpcId);
+      logger.log(Level.FINEST, "Writing Cancel event to GcpLogSink");
+      sink.write(logEntryBuilder.build());
+    }
+  }
+
+  // Copied from internal
+  static final class ByteArrayMarshaller implements Marshaller<byte[]> {
+    @Override
+    public InputStream stream(byte[] value) {
+      return new ByteArrayInputStream(value);
+    }
+
+    @Override
+    public byte[] parse(InputStream stream) {
+      try {
+        return parseHelper(stream);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private byte[] parseHelper(InputStream stream) throws IOException {
+      try {
+        return IoUtils.toByteArray(stream);
+      } finally {
+        stream.close();
+      }
+    }
+  }
+
+  // Copied from internal
+  static final class IoUtils {
+    /** maximum buffer to be read is 16 KB. */
+    private static final int MAX_BUFFER_LENGTH = 16384;
+
+    /** Returns the byte array. */
+    public static byte[] toByteArray(InputStream in) throws IOException {
+      ByteArrayOutputStream out = new ByteArrayOutputStream();
+      copy(in, out);
+      return out.toByteArray();
+    }
+
+    /** Copies the data from input stream to output stream. */
+    public static long copy(InputStream from, OutputStream to) throws IOException {
+      // Copied from guava com.google.common.io.ByteStreams because its API is unstable (beta)
+      checkNotNull(from);
+      checkNotNull(to);
+      byte[] buf = new byte[MAX_BUFFER_LENGTH];
+      long total = 0;
+      while (true) {
+        int r = from.read(buf);
+        if (r == -1) {
+          break;
+        }
+        to.write(buf, 0, r);
+        total += r;
+      }
+      return total;
+    }
+  }
+
+
+  static final class PayloadBuilder<T> {
+    T proto;
+    int size;
+
+    private PayloadBuilder(T proto, int size) {
+      this.proto = proto;
+      this.size = size;
+    }
+  }
+
+  static PayloadBuilder<GrpcLogRecord.Metadata.Builder> createMetadataProto(Metadata metadata) {
```

Comment:
This method is sufficiently complex that it will be useful to have a unit test just to test this method. e.g. you pass metadata with 3 entries (an ascii and a binary value and another one with repeated name). And from the returned `GrpcLogRecord.Metadata.Builder` you build a `GrpcLogRecord.Metadata` and verify its contents : you should find all 3 keys with associated values (with known byte conversion for all 3 metadata values)

Reply:
It is getting tested indirectly as part of events like `REQUEST_HEADER, RESPONSE_HEADER and TRAILER`. 
Added a TODO to add a unit test as well